SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp

```diff
@@ -197,3 +169,121 @@ Worker_ComponentId USpatialDispatcher::GetComponentId(Worker_Op* Op) const
+	{
+		Callback(Op->remove_component);
+	});
+}
+
+USpatialDispatcher::FCallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const TFunction<void(const Worker_AuthorityChangeOp&)>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_AUTHORITY_CHANGE, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->authority_change);
+	});
+}
+USpatialDispatcher::FCallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const TFunction<void(const Worker_ComponentUpdateOp&)>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMPONENT_UPDATE, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->component_update);
+	});
+}
+
+USpatialDispatcher::FCallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const TFunction<void(const Worker_CommandRequestOp&)>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMMAND_REQUEST, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->command_request);
+	});
+}
+
+USpatialDispatcher::FCallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const TFunction<void(const Worker_CommandResponseOp&)>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMMAND_RESPONSE, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->command_response);
+	});
+}
+
+USpatialDispatcher::FCallbackId USpatialDispatcher::AddGenericOpCallback(Worker_ComponentId ComponentId, Worker_OpType OpType, const TFunction<void(const Worker_Op*)>& Callback)
+{
+	check(SpatialConstants::MIN_EXTERNAL_SCHEMA_ID <= ComponentId && ComponentId <= SpatialConstants::MAX_EXTERNAL_SCHEMA_ID);
+	const FCallbackId NewCallbackId = NextCallbackId++;
+	ComponentOpTypeToCallbackMap.FindOrAdd(ComponentId).FindOrAdd(OpType).Add(UserOpCallbackData{ NewCallbackId, Callback });
+	CallbackIdToDataMap.Add(NewCallbackId, CallbackIdData{ ComponentId, OpType });
+	return NewCallbackId;
+}
+
+bool USpatialDispatcher::RemoveOpCallback(FCallbackId CallbackId)
+{
+	// Find callback ID in map and assert if it does not exist
+	CallbackIdData CallbackDataToRemove = CallbackIdToDataMap.FindAndRemoveChecked(CallbackId);
+
+	TMap<Worker_OpType, TArray<UserOpCallbackData>>* OpTypesToCallbacks = ComponentOpTypeToCallbackMap.Find(CallbackDataToRemove.ComponentId);
+	if (OpTypesToCallbacks == nullptr)
+	{
+		return false;
+	}
+
+	TArray<UserOpCallbackData>* ComponentCallbacks = OpTypesToCallbacks->Find(CallbackDataToRemove.OpType);
+	if (ComponentCallbacks == nullptr)
+	{
+		return false;
+	}
+
+	int32 CallbackIndex = ComponentCallbacks->IndexOfByPredicate([CallbackId](const UserOpCallbackData& Data)
+	{
+		return Data.Id == CallbackId;
+	});
+	if (CallbackIndex == INDEX_NONE)
+	{
+		return false;
+	}
+
+	// If removing the only callback for a component ID / op type, delete map entries as applicable
+	if (OpTypesToCallbacks->Num() == 1)
```

Comment:
This seems complicated (and maybe wrong). If I had two callbacks registered for the same component and same op type and wanted to remove just one of them, wouldn't this pass both these checks and remove from the top-level map?

Reply:
ah yep, fucked this up during naming, repushed what I think is correct?